tests/integration/multi_table/test_hma.py

```diff
@@ -1988,3 +1989,112 @@ def test_hma_synthesizer_with_fixed_combinations():
+    """Test the estimated number of columns is correct for various sdtypes.
+
+    To check that the number columns is correct we Mock the ``_finalize`` method
+    and compare its output with the estimated number of columns.
+
+    The dataset used follows the structure below:
+        R1 R2
+        | /
+        GP
+        |
+        P
+    """
+    # Setup
+    root1 = pd.DataFrame({'R1': [0, 1, 2]})
+    root2 = pd.DataFrame({'R2': [0, 1, 2], 'data': [0, 1, 2]})
+    grandparent = pd.DataFrame({'GP': [0, 1, 2], 'R1': [0, 1, 2], 'R2': [0, 1, 2]})
+    parent = pd.DataFrame({
+        'P': [0, 1, 2],
+        'GP': [0, 1, 2],
+        'numerical': [0.1, 0.5, np.nan],
+        'categorical': ['a', np.nan, 'c'],
+        'datetime': [None, '2019-01-02', '2019-01-03'],
+        'boolean': [float('nan'), False, True],
+        'id': [0, 1, 2],
+    })
+    data = {
+        'root1': root1,
+        'root2': root2,
+        'grandparent': grandparent,
+        'parent': parent,
+    }
+    metadata = MultiTableMetadata.load_from_dict({
+        'tables': {
+            'root1': {
+                'primary_key': 'R1',
+                'columns': {
+                    'R1': {'sdtype': 'id'},
+                },
+            },
+            'root2': {
+                'primary_key': 'R2',
+                'columns': {'R2': {'sdtype': 'id'}, 'data': {'sdtype': 'numerical'}},
+            },
+            'grandparent': {
+                'primary_key': 'GP',
+                'columns': {
+                    'GP': {'sdtype': 'id'},
+                    'R1': {'sdtype': 'id'},
+                    'R2': {'sdtype': 'id'},
+                },
+            },
+            'parent': {
+                'primary_key': 'P',
+                'columns': {
+                    'P': {'sdtype': 'id'},
+                    'GP': {'sdtype': 'id'},
+                    'numerical': {'sdtype': 'numerical'},
+                    'categorical': {'sdtype': 'categorical'},
+                    'datetime': {'sdtype': 'datetime'},
+                    'boolean': {'sdtype': 'boolean'},
+                    'id': {'sdtype': 'id'},
+                },
+            },
+        },
+        'relationships': [
+            {
+                'parent_table_name': 'root1',
+                'parent_primary_key': 'R1',
+                'child_table_name': 'grandparent',
+                'child_foreign_key': 'R1',
+            },
+            {
+                'parent_table_name': 'root2',
+                'parent_primary_key': 'R2',
+                'child_table_name': 'grandparent',
+                'child_foreign_key': 'R2',
+            },
+            {
+                'parent_table_name': 'grandparent',
+                'parent_primary_key': 'GP',
+                'child_table_name': 'parent',
+                'child_foreign_key': 'GP',
+            },
+        ],
+    })
+    synthesizer = HMASynthesizer(metadata)
+    synthesizer._finalize = Mock(return_value=data)
```

Comment:
@amontanez24 Are we allowed to use mocks in integration tests?

Reply:
Yeah that's a good question. Ideally you wouldn't have to. This test is a little strange in that it is mocking the internals of another method (_finalize), to assert that the method under test is returning the right values. I'm not really sure of the context, so I don't know why this is the case.

Ideally, this test would just run _estimate_num_columns, and we'd know what the result should be and could just assert that outcome.